### PR TITLE
[FLINK-33676] Implement RestoreTests for WindowAggregate

### PR DESCRIPTION
**Pull request by jnh5y**
## What is the purpose of the change

Implement restore tests for WindowAggregate node.

## Verifying this change

This change added tests and can be verified as follows:

* Added restore tests for WindowAggregate node which verifies the generated compiled plan with the saved compiled plan

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

**Comment by jnh5y** [comment]
@flinkbot run azure

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/stream/WindowAggregateEventTimeRestoreTest.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+public class WindowAggregateEventTimeRestoreTest extends RestoreTestBase {
+
+    public WindowAggregateEventTimeRestoreTest() {
+        super(StreamExecWindowAggregate.class);
+    }
+
+    @Override
+    public List<TableTestProgram> programs() {
+        return Arrays.asList(
+                WindowAggregateTestPrograms.GROUP_TUMBLE_WINDOW_EVENT_TIME,
```
Comment (xuyangzhong) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/stream/WindowAggregateEventTimeRestoreTest.java:37]: Just wonder why add prefix 'GROUP_' each test?

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/stream/WindowAggregateTestPrograms.java [hunk 2]
```diff
@@ -0,0 +1,494 @@
...
+                                            "+I[a, 1, 10, 1]", "+I[c, 1, 12, 1]", "+I[d, 2, 24, 2]")
+                                    .build())
+                    .runSql(
+                            "INSERT INTO sink_t SELECT "
+                                    + "name, "
+                                    + "COUNT(*), "
+                                    + "SUM(a_int), "
+                                    + "COUNT(DISTINCT `comment`) "
+                                    + "FROM source_t "
+                                    + "GROUP BY name, TUMBLE(proctime, INTERVAL '5' SECOND)")
```
Comment (xuyangzhong) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/stream/WindowAggregateTestPrograms.java:448]: All tests in this class are about window agg, but this test is related about group window agg, right?

**Review by xuyangzhong — COMMENTED** [review]
Hi, thanks for this driving. I just left some minor comments.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/stream/WindowAggregateTestPrograms.java (hunk 2 above)
Comment (jnh5y) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/stream/WindowAggregateTestPrograms.java:449]: Yes!  Good catch.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/stream/WindowAggregateEventTimeRestoreTest.java (hunk 1 above)
Comment (jnh5y) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/stream/WindowAggregateEventTimeRestoreTest.java:37]: Ah, this was a mistake on my part.  Corrected.

**Comment by jnh5y** [comment]
@flinkbot run azure

**Review by dawidwys — CHANGES_REQUESTED** [review]

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/stream/WindowAggregateTestPrograms.java [hunk 3]
```diff
@@ -0,0 +1,528 @@
...
+import org.apache.flink.table.test.program.TableTestProgram;
+import org.apache.flink.types.Row;
+
+import java.math.BigDecimal;
+import java.util.function.Function;
+
+/** {@link TableTestProgram} definitions for testing {@link StreamExecWindowAggregate}. */
+public class WindowAggregateTestPrograms {
+
+    static final Row[] BEFORE_DATA = {
```
[the remark this answers is not in the document]
Comment (dawidwys) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/stream/WindowAggregateTestPrograms.java:34]: Generally, could you go over all the visibility modifiers and set the minimal required one.

**Comment by dawidwys** [comment]
Thank you for the PR @jnh5y I left some comments.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/stream/WindowAggregateTestPrograms.java [hunk 4]
```diff
@@ -0,0 +1,528 @@
...
+                            .addOption("changelog-mode", "I,UA,UB,D")
+                            .producedBeforeRestore(BEFORE_DATA)
+                            .producedAfterRestore(AFTER_DATA);
+    static final SourceTestStep SOURCE = SOURCE_BUILDER.apply("window_source_t").build();
+
+    static final SourceTestStep CDC_SOURCE =
+            SOURCE_BUILDER
+                    .apply("cdc_window_source_t")
+                    .addOption("changelog-mode", "I,UA,UB,D")
+                    .build();
```
[the remark this answers is not in the document]
Comment (jnh5y) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/stream/WindowAggregateTestPrograms.java:80]: Ah, I made a mistake; removing the changelog for the SOURCE_BUILDER.

Just to say it out loud, the CDC_SOURCE is set up to help cover some of the tests which @xuyangzhong added like this: https://github.com/apache/flink/blob/master/flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/stream/WindowAggregateJsonPlanTest.java#L98

Overall, there are several orthogonal options, and at some point, we should only test a subset of them.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/stream/WindowAggregateTestPrograms.java [hunk 5]
```diff
@@ -0,0 +1,528 @@
...
+                    "SESSION(TABLE window_source_t PARTITION BY name, DESCRIPTOR(rowtime), INTERVAL '5' SECOND)",
+                    SESSION_EVENT_TIME_PARTITIONED_BEFORE_ROWS,
+                    SESSION_EVENT_TIME_PARTITIONED_AFTER_ROWS);
+
+    static final TableTestProgram SESSION_WINDOW_PARTITION_EVENT_TIME_TWO_PHASE_DISTINCT_SPLIT =
+            getTableTestProgram(
+                    "window-aggregate-session-partition-event-time-two-phase-distinct-split",
+                    "validates group by using session window with event time with two phase aggregation",
+                    AggregatePhaseStrategy.TWO_PHASE.toString(),
+                    "SESSION(TABLE cdc_window_source_t PARTITION BY name, DESCRIPTOR(rowtime), INTERVAL '5' SECOND)",
```
[the remark this answers is not in the document]
Comment (jnh5y) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/stream/WindowAggregateTestPrograms.java:466]: Answered above.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/stream/WindowAggregateTestPrograms.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (jnh5y) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/stream/WindowAggregateTestPrograms.java:34]: Yes.  Doing it.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/stream/WindowAggregateTestPrograms.java [hunk 6]
```diff
@@ -0,0 +1,528 @@
...
+        "-U[b, 2020-10-10T00:00:01, 2020-10-10T00:00:13, 1, 3, 1]",
+        "+U[b, 2020-10-10T00:00:01, 2020-10-10T00:00:13, 2, 9, 2]",
+        "-U[a, 2020-10-10T00:00:01, 2020-10-10T00:00:13, 4, 10, 2]",
+        "+U[a, 2020-10-10T00:00:01, 2020-10-10T00:00:13, 5, 13, 3]",
+        "-U[a, 2020-10-10T00:00:01, 2020-10-10T00:00:13, 5, 13, 3]",
+        "+U[a, 2020-10-10T00:00:01, 2020-10-10T00:00:13, 6, 18, 3]",
+        "+I[b, 2020-10-10T00:00:16, 2020-10-10T00:00:21, 1, 4, 1]"
+    };
+
+    public static final String[] SESSION_EVENT_TIME_AFTER_ROWS = {
```
[the remark this answers is not in the document]
Comment (jnh5y) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/stream/WindowAggregateTestPrograms.java:391]: Switching to package.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/stream/WindowAggregateJsonPlanTest.java [hunk 7]
```diff
@@ -1,528 +0,0 @@
...
-                "insert into MySink select\n"
-                        + "  window_end,\n"
-                        + "  COUNT(*)\n"
-                        + "FROM TABLE(\n"
-                        + "   SESSION(TABLE MyTable, DESCRIPTOR(proctime), INTERVAL '15' MINUTE))\n"
-                        + "GROUP BY window_start, window_end");
-    }
-
-    @Test
-    void testProcTimeSessionWindowWithPartitionKey() {
```
[the remark this answers is not in the document]
Comment (jnh5y) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/stream/WindowAggregateJsonPlanTest.java:472]: I think so.  These tests should cover it.

**Comment by jnh5y** [comment]
@flinkbot run azure

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/stream/WindowAggregateJsonPlanTest.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (dawidwys) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/stream/WindowAggregateJsonPlanTest.java:472]: I think I put the comment on a wrong line. I see we test in `WindowAggregateJsonPlanTest`:
1. `SESSION(TABLE MyTable PARTITION BY b, DESCRIPTOR(rowtime), INTERVAL '15' MINUTE))`
2. `SESSION(TABLE MyTable, DESCRIPTOR(rowtime), INTERVAL '15' MINUTE)

I couldn't find 2 in new tests.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/stream/WindowAggregateTestPrograms.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (dawidwys) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/stream/WindowAggregateTestPrograms.java]: Ok, so why cannot this field be `private` ?

[the remark this answers is not in the document]
Comment (dawidwys) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/stream/WindowAggregateTestPrograms.java]: Have you actually checked the modifiers? The majority of fields are still in the `default` scope. I don't see a reason why they could not be private.

There are still `public` fields that I believe could be `private`, e.g. `TUMBLE_EVENT_TIME_AFTER_ROWS`. Or is it used elsewhere?

**Review by dawidwys — COMMENTED** [review]
Thank you for the update! It looks almost good. I'd like to clarify the partitioning cases and after that should be good 👍

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/stream/WindowAggregateJsonPlanTest.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (jnh5y) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/stream/WindowAggregateJsonPlanTest.java:472]: I think number 2 is covered by https://github.com/apache/flink/pull/23886/files#diff-71e15316241988f3793452ab51740da6880659d3e5c8e252ba1e8554ba13defeR410

`"SESSION(TABLE window_source_t, DESCRIPTOR(rowtime), INTERVAL '5' SECOND)",`

I will admit that I may have skipped some of the combinations... with 4 types of windows, three or so options (two phase, two phase distinct, one phase), and offsets, the number of combinations got a little bit too big.  

I think I've covered all of the components of the compiled plan though.

**Pull request closed without merging**